framework/src/userobject/ElementQualityChecker.C

```diff
@@ -0,0 +1,201 @@
+  {
+    if (_upper_bound > default_bounds.first)
+      actual_bounds = std::make_pair(default_bounds.first, _upper_bound);
+    else
+      mooseError("Provided upper bound should larger than the default lower bound: ",
+                 default_bounds.first);
+  }
+  else
+    actual_bounds = default_bounds;
+
+  // calculate and save quality metric value for current element
+  Real mv = _current_elem->quality(_m_type);
+  _m_values.insert(mv);
+
+  // check element quality metric, save ids of elements whose quality metrics exceeds the preset
+  // bounds
+  if (mv < actual_bounds.first || mv > actual_bounds.second)
+    _elem_ids.insert(_current_elem->id());
+}
+
+void
+ElementQualityChecker::threadJoin(const UserObject & uo)
+{
+  const ElementQualityChecker & eqc = static_cast<const ElementQualityChecker &>(uo);
+  _m_values.insert(eqc._m_values.begin(), eqc._m_values.end());
+  _elem_ids.insert(eqc._elem_ids.begin(), eqc._elem_ids.end());
+  _bypassed += eqc._bypassed;
+}
+
+void
+ElementQualityChecker::finalize()
+{
+  _communicator.set_union(_m_values);
+  _communicator.set_union(_elem_ids);
+  _communicator.sum(_bypassed);
+
+  if (_bypassed)
+  {
+    switch (_failure_type)
+    {
+      case FailureType::WARNING:
+        mooseWarning("Provided quality metric doesn't apply to certain element type!");
+        break;
+
+      case FailureType::ERROR:
+        mooseError("Provided quality metric doesn't apply to certain element type!");
+        break;
+
+      default:
+        mooseError("Unknown failure type!");
+    }
+  }
+
+  Real m_sum = 0;
+  for (std::set<Real>::iterator it = _m_values.begin(); it != _m_values.end(); ++it)
+    m_sum += *it;
+
+  Moose::out << libMesh::Quality::name(_m_type) << " Metric values:" << std::endl;
+  Moose::out << "              Minimum: " << *_m_values.begin() << std::endl;
+  Moose::out << "              Maximum: " << *_m_values.rbegin() << std::endl;
+  Moose::out << "              Average: " << m_sum / _m_values.size() << std::endl;
+
+  if (!_elem_ids.empty())
+  {
+    switch (_failure_type)
+    {
+      case FailureType::WARNING:
+      {
+        mooseWarning("WARNING: Bad element quality in terms of ",
+                     libMesh::Quality::name(_m_type),
+                     " quality metric for ",
+                     _elem_ids.size(),
+                     " elements.");
+        break;
+      }
+
+      case FailureType::ERROR:
+      {
+        mooseError("ERROR: Bad element quality in terms of ",
+                   libMesh::Quality::name(_m_type),
+                   " quality metric for ",
+                   _elem_ids.size(),
+                   " elements.");
+        break;
+      }
+
+      default:
+        mooseError("Unknown failure type!");
+    }
+
+    Moose::out << "List of failed element IDs: " << std::endl;
+    for (std::set<unsigned int>::iterator it = _elem_ids.begin(); it != _elem_ids.end(); ++it)
+      Moose::out << "  " << *it << std::endl;
```

Comment:
We have a `stringify` utility for printing containers out. You might want to try it.